helix-core/src/test/java/org/apache/helix/integration/task/TestWorkflowContextWithoutConfig.java

```diff
@@ -0,0 +1,131 @@
+import org.apache.helix.model.IdealState;
+import org.apache.helix.task.JobConfig;
+import org.apache.helix.task.TaskState;
+import org.apache.helix.task.Workflow;
+import org.apache.helix.task.WorkflowConfig;
+import org.apache.helix.task.WorkflowContext;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * Test to check workflow context is not created without workflow config.
+ */
+public class TestWorkflowContextWithoutConfig extends TaskTestBase {
+  private HelixAdmin _admin;
+
+  @BeforeClass
+  public void beforeClass() throws Exception {
+    super.beforeClass();
+    _admin = _gSetupTool.getClusterManagementTool();
+  }
+
+  @Test
+  public void testWorkflowContextWithoutConfig() throws Exception {
+    final long expiryTime = 5000L;
+    String workflowName1 = TestHelper.getTestMethodName() + "_1";
+    Workflow.Builder builder1 = new Workflow.Builder(workflowName1);
+
+    // Workflow DAG Schematic:
+    //          JOB0
+    //           /\
+    //          /  \
+    //         /    \
+    //       JOB1   JOB2
+
+    JobConfig.Builder jobBuilder1 = JobConfig.Builder.fromMap(WorkflowGenerator.DEFAULT_JOB_CONFIG)
+        .setMaxAttemptsPerTask(1).setWorkflow(workflowName1)
+        .setJobCommandConfigMap(ImmutableMap.of(MockTask.JOB_DELAY, "1000"));
+
+    JobConfig.Builder jobBuilder2 = JobConfig.Builder.fromMap(WorkflowGenerator.DEFAULT_JOB_CONFIG)
+        .setMaxAttemptsPerTask(1).setWorkflow(workflowName1)
+        .setJobCommandConfigMap(ImmutableMap.of(MockTask.JOB_DELAY, "1000"));
+
+    JobConfig.Builder jobBuilder3 = JobConfig.Builder.fromMap(WorkflowGenerator.DEFAULT_JOB_CONFIG)
+        .setMaxAttemptsPerTask(1).setWorkflow(workflowName1)
+        .setJobCommandConfigMap(ImmutableMap.of(MockTask.JOB_DELAY, "1000"));
+
+    builder1.addParentChildDependency("JOB0", "JOB1");
+    builder1.addParentChildDependency("JOB0", "JOB2");
+    builder1.addJob("JOB0", jobBuilder1);
+    builder1.addJob("JOB1", jobBuilder2);
+    builder1.addJob("JOB2", jobBuilder3);
+    builder1.setExpiry(expiryTime);
+
+    _driver.start(builder1.build());
+
+    // Wait until workflow is created and IN_PROGRESS state.
+    _driver.pollForWorkflowState(workflowName1, TaskState.IN_PROGRESS);
+
+    // Check that WorkflowConfig, WorkflowContext, and IdealState are indeed created for this
+    // workflow
+    Assert.assertNotNull(_driver.getWorkflowConfig(workflowName1));
+    Assert.assertNotNull(_driver.getWorkflowContext(workflowName1));
+    Assert.assertNotNull(_admin.getResourceIdealState(CLUSTER_NAME, workflowName1));
+
+    // Wait until workflow is completed.
+    _driver.pollForWorkflowState(workflowName1, TaskState.COMPLETED);
+
+    String idealStatePath = "/" + CLUSTER_NAME + "/IDEALSTATES/" + workflowName1;
+    ZNRecord record = _manager.getHelixDataAccessor().getBaseDataAccessor().get(idealStatePath,
```

Comment:
Could this possibly be null?

Reply:
Idealstate will be deleted (null) when workflow expired. If the workflow is in progress, IdealState should not be null. Just to make sure, I moved this part before polling for COMPLETED state. Also I added an assert to make sure record is not null. Thanks for the comment.

Reply:
Doesn't the idealstate for the workflow get dropped after it completed? Or am I thinking of jobs' IdealStates?

Reply:
The idealstate of the job will be dropped when job is failed or completed. For the workflow idealstate is deleted when workflow is expired.